cloudbridge/cloud/factory.py

```diff
@@ -12,6 +16,54 @@
 log = logging.getLogger(__name__)
 
 
+# Todo: Move to pyeventsystem if we're keeping this logic
+class ParentMiddlewareManager(SimpleMiddlewareManager):
+
+    def __init__(self, event_manager=None):
+        super(ParentMiddlewareManager, self).__init__(event_manager)
+        self.middleware_constructors = []
+
+    def add_constructor(self, middleware_class, *args):
+        self.middleware_constructors.append((middleware_class, args))
+
+    def remove_constructor(self, middleware_class, *args):
+        self.middleware_constructors.remove((middleware_class, args))
+
+    def generate_simple_manager(self):
+        new_manager = SimpleMiddlewareManager()
+        for middleware in self.middleware_list:
+            new_manager.add(middleware)
+        for constructor, args in self.middleware_constructors:
+            m = constructor(*args)
+            new_manager.add(m)
+        for handler in self.get_subscribed_handlers():
+            new_handler = handler.__class__(handler.event_pattern,
+                                            handler.priority,
+                                            handler.callback)
+            new_manager.events.subscribe(new_handler)
+        return new_manager
+
+    # Removing install step from add. Since this manager is meant to create
+    # other managers rather than run. This will also simplify separating
+    # handlers added through middleware from those subscribed directly
+    def add(self, middleware):
```

Comment:
I don't think we should provide dual semantics here  - one for shared middleware objects, and another for independent middleware, just to keep it simple. I'm currently leaning in favour of the factory namespacing child event dispatchers, and forwarding them as necessary, so that we still maintain event isolation within a given provider. I think we should avoid a scenario where providers are unwittingly subscribing to events outside of the provider.